// telegram-chat-parser/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentType = exports.ChatType = exports.TelegramUser = exports.TelegramMessage = exports.TelegramChat = exports.Types = void 0;
/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable import/no-cycle */
/* eslint-disable import/prefer-default-export */
var types = __importStar(require("./types"));
var Types;
(function (Types) {
    Types.mod = types;
})(Types = exports.Types || (exports.Types = {}));
var TelegramChat_1 = require("./TelegramChat");
Object.defineProperty(exports, "TelegramChat", { enumerable: true, get: function () { return __importDefault(TelegramChat_1).default; } });
var TelegramMessage_1 = require("./TelegramMessage");
Object.defineProperty(exports, "TelegramMessage", { enumerable: true, get: function () { return __importDefault(TelegramMessage_1).default; } });
var TelegramUser_1 = require("./TelegramUser");
Object.defineProperty(exports, "TelegramUser", { enumerable: true, get: function () { return __importDefault(TelegramUser_1).default; } });
var ChatType_1 = require("./ChatType");
Object.defineProperty(exports, "ChatType", { enumerable: true, get: function () { return __importDefault(ChatType_1).default; } });
var ContentType_1 = require("./ContentType");
Object.defineProperty(exports, "ContentType", { enumerable: true, get: function () { return __importDefault(ContentType_1).default; } });
